File: web/lib/markdown.ts
import { Cell, CellType } from '../components/CellList';
import { v4 as uuidv4 } from 'uuid';

export function parseMarkdown(content: string): Cell[] {
  const cells: Cell[] = [];
  let currentText = '';
  const lines = content.split('\n');
  let i = 0;

  while (i < lines.length) {
    const line = lines[i];

    if (line.trimStart().startsWith('```')) {
      // Flush current text as Markdown cell
      if (currentText.trim().length > 0) {
        const id = uuidv4();
        // Remove leading newline if present
        let cellContent = currentText.startsWith('\n') ? currentText.slice(1) : currentText;
        cellContent = cellContent.trimEnd();

        cells.push({
          id,
          content: cellContent,
          output: '',
          type: 'markdown'
        });
        currentText = '';
      }

      // Parse Code Block
      const langTag = line.trimStart().replace('```', '').trim();

      const cellType: CellType = (() => {
        switch (langTag) {
          case 'rust': return 'rust';
          case 'python':
          case 'py': return 'python';
          case 'javascript':
          case 'js': return 'javascript';
          case 'typescript':
          case 'ts': return 'typescript';
          case 'c': return 'c';
          case 'cpp':
          case 'c++': return 'cpp';
          case 'go': return 'go';
          case 'bash':
          case 'sh':
          case 'shell': return 'shell';
          case 'markdown':
          case 'md': return 'markdown';
          default: return 'shell'; // if codeblock with no specific tag then it's shell
        }
      })();

      // Collect code content
      let codeContent = '';
      i++; // Move past opening fence
      while (i < lines.length) {
        if (lines[i].trimStart().startsWith('```')) {
          i++; // Consume closing fence
          break;
        }
        codeContent += lines[i] + '\n';
        i++;
      }

      // Parse Output (lines starting with ">> ")
      let outputContent = '';
      while (i < lines.length && lines[i].startsWith('>> ')) {
        outputContent += lines[i].slice(3) + '\n';
        i++;
      }

      const id = uuidv4();

      if (cellType === 'markdown') {
        cells.push({
          id,
          content: codeContent.trimEnd(),
          output: '',
          type: cellType
        });
      } else {
        cells.push({
          id,
          content: codeContent.trimEnd(),
          output: outputContent.trimEnd(),
          type: cellType
        });
      }
    } else {
      currentText += line + '\n';
      i++;
    }
  }

  // Flush remaining text
  if (currentText.trim().length > 0) {
    const id = uuidv4();
    // Remove leading newline if present
    let cellContent = currentText.startsWith('\n') ? currentText.slice(1) : currentText;
    cellContent = cellContent.trimEnd();

    cells.push({
      id,
      content: cellContent,
      output: '',
      type: 'markdown'
    });
  }

  // Ensure at least one cell
  if (cells.length === 0) {
    cells.push({
      id: uuidv4(),
      content: '',
      output: '',
      type: 'shell'
    });
  }

  return cells;
}

export function toMarkdown(cells: Cell[]): string {
  let output = '';

  for (let i = 0; i < cells.length; i++) {
    const cell = cells[i];

    if (i > 0) {
      output += '\n';
    }

    if (cell.type === 'markdown') {
      output += cell.content;
      output += '\n';
    } else {
      const lang: string = (() => {
        const t = cell.type;
        if (t === 'rust') return 'rust';
        if (t === 'python') return 'python';
        if (t === 'javascript') return 'javascript';
        if (t === 'typescript') return 'typescript';
        if (t === 'c') return 'c';
        if (t === 'cpp') return 'cpp';
        if (t === 'go') return 'go';
        if (t === 'shell') return 'bash';
        return 'bash';
      })();

      output += `\`\`\`${lang}\n`;
      output += cell.content;
      if (!cell.content.endsWith('\n')) {
        output += '\n';
      }
      output += '```\n';

      if (cell.output && cell.output.length > 0) {
        for (const line of cell.output.split('\n')) {
          output += `>> ${line}\n`;
        }
      }
    }
  }

  return output;
}

File: term/src/markdown.rs
use crate::{Cell, CellType};

pub fn parse_markdown(content: &str) -> Vec<Cell> {
    let mut cells = Vec::new();
    let mut current_text = String::new();
    let mut lines = content.lines().peekable();

    while let Some(line) = lines.next() {
        if line.trim_start().starts_with("```") {
            // Flush current text as Markdown cell
            if !current_text.trim().is_empty() {
                 let id = uuid::Uuid::new_v4().to_string();
                 // Remove leading newline if present
                 let content = if current_text.starts_with('\n') {
                     &current_text[1..]
                 } else {
                     &current_text
                 };
                 cells.push(Cell {
                     id,
                     content: content.trim_end().to_string(),
                     output: String::new(),
                     cell_type: CellType::Markdown,
                     polling_interval: None,
                     last_run: None,
                 });
                 current_text.clear();
            }

            // Parse Code Block
            let lang_tag = line.trim_start().trim_start_matches("```").trim();

            let cell_type = match lang_tag {
                "rust" => CellType::Rust,
                "python" | "py" => CellType::Python,
                "javascript" | "js" => CellType::JavaScript,
                "typescript" | "ts" => CellType::TypeScript,
                "c" => CellType::C,
                "cpp" | "c++" => CellType::Cpp,
                "go" => CellType::Go,
                "bash" | "sh" | "shell" => CellType::Shell,
                _ => CellType::Shell, //if codeblock with no specific tag then it's shell
            };

            let mut code_content = String::new();
            while let Some(code_line) = lines.peek() {
                if code_line.trim_start().starts_with("```") {
                    lines.next(); // Consume closing fence
                    break;
                }
                code_content.push_str(lines.next().unwrap());
                code_content.push('\n');
            }
            
            // Parse Output
            let mut output_content = String::new();
            while let Some(out_line) = lines.peek() {
                if out_line.starts_with(">> ") {
                    output_content.push_str(&out_line[3..]);
                    output_content.push('\n');
                    lines.next();
                } else {
                    break;
                }
            }

            let id = uuid::Uuid::new_v4().to_string();
            
            if cell_type == CellType::Markdown {
                cells.push(Cell {
                    id,
                    content: code_content.trim_end().to_string(),
                    output: String::new(),
                    cell_type,
                    polling_interval: None,
                    last_run: None,
                });
            } else {
                cells.push(Cell {
                    id,
                    content: code_content.trim_end().to_string(),
                    output: output_content.trim_end().to_string(),
                    cell_type,
                    polling_interval: None,
                    last_run: None,
                });
            }

        } else {
            current_text.push_str(line);
            current_text.push('\n');
        }
    }
    
    // Flush remaining text
    if !current_text.trim().is_empty() {
         let id = uuid::Uuid::new_v4().to_string();
         // Remove leading newline if present
         let content = if current_text.starts_with('\n') {
             &current_text[1..]
         } else {
             &current_text
         };
         cells.push(Cell {
             id,
             content: content.trim_end().to_string(),
             output: String::new(),
             cell_type: CellType::Markdown,
             polling_interval: None,
             last_run: None,
         });
    }

    // Ensure at least one cell
    if cells.is_empty() {
        cells.push(Cell {
            id: uuid::Uuid::new_v4().to_string(),
            content: String::new(),
            output: String::new(),
            cell_type: CellType::Shell,
            polling_interval: None,
            last_run: None,
        });
    }

    cells
}

pub fn to_markdown(cells: &[Cell]) -> String {
    let mut output = String::new();
    
    for (i, cell) in cells.iter().enumerate() {
        if i > 0 {
            output.push('\n');
        }

        if cell.cell_type == CellType::Markdown {
            output.push_str(&cell.content);
            output.push('\n');
        } else {
            let lang = match cell.cell_type {
                CellType::Rust => "rust",
                CellType::Python => "python",
                CellType::JavaScript => "javascript",
                CellType::TypeScript => "typescript",
                CellType::C => "c",
                CellType::Cpp => "cpp",
                CellType::Go => "go",
                CellType::Shell => "bash",
                CellType::Markdown => "markdown", 
            };
            
            output.push_str(&format!("```{}\n", lang));
            output.push_str(&cell.content);
            if !cell.content.ends_with('\n') {
                output.push('\n');
            }
            output.push_str("```\n");
            
            if !cell.output.is_empty() {
                for line in cell.output.lines() {
                    output.push_str(&format!(">> {}\n", line));
                }
            }
        }
    }
    
    output
}
